spiffe/tls_verify.go

```diff
@@ -0,0 +1,50 @@
+package spiffe
+
+import (
+	"crypto/x509"
+	"errors"
+
+	"github.com/spiffe/go-spiffe/internal"
+)
+
+// VerifyPeerCertificate verifies the provided peer certificate chain using the
+// set trust domain roots. The expectPeerFn callback is used to check the peer
+// ID after the chain of trust has been verified to assert that the chain
+// belongs to the intended peer.
+func VerifyPeerCertificate(peerChain []*x509.Certificate, trustDomainRoots map[string]*x509.CertPool, expectPeerFn ExpectPeerFunc) ([][]*x509.Certificate, error) {
+	switch {
+	case len(peerChain) == 0:
+		return nil, errors.New("no peer certificates")
+	case len(trustDomainRoots) == 0:
+		return nil, errors.New("at least one trust domain root is required")
+	case expectPeerFn == nil:
+		return nil, errors.New("expectPeerFn callback is required")
+	}
+
+	peer := peerChain[0]
+	peerID, trustDomainID, err := internal.GetIDsFromCertificate(peer)
+	if err != nil {
+		return nil, err
+	}
+
+	roots, ok := trustDomainRoots[trustDomainID]
+	if !ok {
+		return nil, errors.New("no roots for peer trust domain")
```

Comment:
Perhaps include the trust domain ID/name here?

Reply:
:+1: